extensions/cornerstone/src/init.js

```diff
@@ -1,40 +1,239 @@
 import initCornerstoneTools from './initCornerstoneTools.js';
-import cornerstoneTools from 'cornerstone-tools';
-import initWADOImageLoader from './initWADOImageLoader.js';
+import './initWADOImageLoader.js';
+import getCornerstoneMeasurementById from './utils/getCornerstoneMeasurementById';
 import measurementServiceMappingsFactory from './utils/measurementServiceMappings/measurementServiceMappingsFactory';
-//
 import { setEnabledElement } from './state';
 
+// TODO -> Global "context menu open state", or lots of expensive searches on drag?
+
+let CONTEXT_MENU_OPEN = false;
+
+const { globalImageIdSpecificToolStateManager } = csTools;
+
+const TOOL_TYPES_WITH_CONTEXT_MENU = [
+  'Angle',
+  'ArrowAnnotate',
+  'Bidirectional',
+  'Length',
+  'FreehandMouse',
+  'EllipticalRoi',
+  'CircleRoi',
+  'RectangleRoi',
+];
+
+const _refreshViewport = () =>
+  cs.getEnabledElements().forEach(({ element }) => cs.updateImage(element));
+
 /**
  *
  * @param {Object} servicesManager
  * @param {Object} configuration
  * @param {Object|Array} configuration.csToolsConfig
  */
-export default function init({ servicesManager, configuration }) {
+export default function init({
+  servicesManager,
+  commandsManager,
+  configuration,
+}) {
   const {
     UIDialogService,
     MeasurementService,
     DisplaySetService,
   } = servicesManager.services;
 
-  const callInputDialog = (data, event, callback) => {
+  /* Measurement Service */
+  const measurementServiceSource = _connectToolsToMeasurementService(
+    MeasurementService,
+    DisplaySetService
+  );
+
+  const onRightClick = event => {
+    if (!UIDialogService) {
+      console.warn('Unable to show dialog; no UI Dialog Service available.');
+      return;
+    }
+
+    const onGetMenuItems = defaultMenuItems => {
+      const { element, currentPoints } = event.detail;
+      const nearbyToolData = commandsManager.runCommand('getNearbyToolData', {
+        element,
+        canvasCoordinates: currentPoints.canvas,
+        availableToolTypes: TOOL_TYPES_WITH_CONTEXT_MENU,
+      });
+
+      let menuItems = [];
+      if (nearbyToolData) {
+        defaultMenuItems.forEach(item => {
+          item.value = nearbyToolData;
+          menuItems.push(item);
+        });
+      }
+
+      return menuItems;
+    };
+
+    CONTEXT_MENU_OPEN = true;
+
+    UIDialogService.dismiss({ id: 'context-menu' });
+    UIDialogService.create({
+      id: 'context-menu',
+      isDraggable: false,
+      preservePosition: false,
+      defaultPosition: _getDefaultPosition(event.detail),
+      content: ContextMenuMeasurements,
+      onClickOutside: () => {
+        UIDialogService.dismiss({ id: 'context-menu' });
+        CONTEXT_MENU_OPEN = false;
+      },
+      contentProps: {
+        onGetMenuItems,
+        eventData: event.detail,
+        onDelete: item => {
+          const { tool: measurementData, toolType } = item.value;
+          measurementServiceSource.remove(measurementData.id);
```

Comment:
do we need this AND commandsManager.runCommand('removeToolState') which happens immediately afterwards?

Reply:
fixed.